### DISCUSSION TITLE
Учёт инструмента выданного персоналу. Настройка подчинённых таблиц.

### DISCUSSION THREAD

**Discussion by EKhorikov** [totumonline/totum-ru-issues-and-discussions#28] [Как это сделать?] [answered]
Друзья, всем привет.

Разрабатываю, для себя, в totum полезные в работе модули.
 
Во первых, хочу сказать - очень круто, что появился такой сервис как totum, очень крутая штука, в которой есть возможность разрабатывать самостоятельно, без привлечения программистов. Спасибо Алексею Полунину и команде, за проделанную работу.

В данный момент, начал делать простенький модуль, для учёта персонала и выданного для работы инструмента и имущества компании. С самими таблицами вопросов нет, а вот с подчинёнными проблема. Делаю всё сам, хоть и не программист, поэтому буду благодарен за подробные объяснения)).

**И так обо всём по порядку и ближе к сути вопроса:**

Есть таблица циклов, со списком сотрудников, в которой учитываются основные данные сотрудников, учитываемые при принятии на работу. Расчётных таблиц (вкладок) в цикле две, вторая таблица, это "имущество", это  и есть таблица, которая призвана учитывать выданный во временное пользование сотруднику, инструмент, инвентарь или любую другую собственность компании.

![image](https://user-images.githubusercontent.com/92679628/137633261-bce97161-8212-4403-bc80-d5894a20c22f.png)

Так же, есть отдельная таблица (циклов или простая, в моём случае циклов, т.к. планируется и в ней сделать несколько расчётных таблиц, для учёта истории изменений например (кому выдавался инструмент ранее)), "Имущество на балансе" которая является сводной, для всего имущества компании. В неё вносится любое имущество, поступившее на баланс компании. Она содержит поля, с информацией о каждой имущественной единицы, а так же, поле СЕЛЕКТ, которое ссылается на список сотрудников, по ФИО, описанный выше.

![image](https://user-images.githubusercontent.com/92679628/137633391-b7da8e58-71c1-41cb-87f5-29259b4f495a.png)

### Задача: 
**Необходимо настроить подчинённость таблиц таким образом, чтоб при выборе конкретного сотрудника, в таблице "Имущество на балансе", в конкретном артикуле, этот артикул, и только он, появлялся во вкладке "Имущество" в таблице "Сотрудники", для учёта имущества выданного каждому сотруднику**

> К слову сказать, подобная стандартная настройка подчинённости таблиц есть в MS Access, там это реализовано очень просто, есть меню, в котором выбираются подчинённые таблицы, которые необходимо отображать в основной, и поля, которые будут являться основными, в нашем случае, это ФИО сотрудника и id.
> 
> Буду очень благодарен за помощь в данном вопросе, важно не просто получить результат, а разобраться в настройке, для применения подобных настроек в дальнейшем. Надеюсь, что мой вопрос будет полезен ещё кому-то.

Спасибо.

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Я чуть упрощу вашу задачу взяв таблицу инструмента как **простую.**

Вот как все выглядит.

Сотрудники - это цикл. Каждый цикл — сотрудник:

![Снимок экрана 2021-10-18 в 15 31 44](https://user-images.githubusercontent.com/55755565/137731765-8aee128b-b2ed-478e-a0e5-07566872e18c.png)

Внутри цикла в моем случае будет одна таблица — в ней имя и блок [автозаполнения строчной части](https://docs.totum.online/autofill-tables):

![Снимок экрана 2021-10-18 в 15 33 53](https://user-images.githubusercontent.com/55755565/137732025-81a4f491-5ff4-45c4-a5c4-90a6124343f3.png)

Те мы берем `list` всех инструментов относящихся к этому циклу (сотруднику).

В инструменте у нас названия инструмена и селект по циклам-сотрудникам:

![Снимок экрана 2021-10-18 в 15 32 33](https://user-images.githubusercontent.com/55755565/137732340-6d91ac30-0400-4587-9dd3-e06c5de023a2.png)

Назначаем инструментам сотрудников. И когда мы заходим в цикл к сотруднику — посчитается `h_insert` и заполнится `insert` в строчной части:

![Снимок экрана 2021-10-18 в 15 42 30](https://user-images.githubusercontent.com/55755565/137732885-85e2f661-30bd-4ee3-86a4-4daf8e08d262.png)
---
![Снимок экрана 2021-10-18 в 15 42 43](https://user-images.githubusercontent.com/55755565/137732911-c42cbadb-31e1-4320-ab41-5c89542bc9a5.png)

Теперь нам нужно добавить код действия на все триггеры в поле `Выдан сотруднику` таблицы `Инструменты`, что бы циклы сотрудников обновлялись в фоне:

![Снимок экрана 2021-10-18 в 15 35 48](https://user-images.githubusercontent.com/55755565/137733220-ccbbd649-84e8-4c38-a03d-fd99243fd630.png)

Те мы по всем триггерам кроме двойного клика пересчитываем циклы, к которым относился инструмент.

+ `$#nfv` — текущее значение.

+ `$#onfv` — предыдущее значение.
